src/RoomDirectoryModel.cpp

```diff
@@ -0,0 +1,185 @@
+
+        server_ = s.toStdString();
+
+        nhlog::ui()->debug("Received matrix server: {}", server_);
+
+        resetDisplayedData();
+
+        endResetModel();
+}
+
+void
+RoomDirectoryModel::setSearchTerm(const QString &f)
+{
+        beginResetModel();
+
+        userSearchString_ = f.toStdString();
+
+        nhlog::ui()->debug("Received user query: {}", userSearchString_);
+
+        resetDisplayedData();
+
+        endResetModel();
+}
+
+std::vector<std::string>
+RoomDirectoryModel::getViasForRoom(const std::vector<std::string> &aliases)
+{
+        std::vector<std::string> vias;
+
+        vias.reserve(aliases.size());
+
+        std::transform(
+          aliases.begin(), aliases.end(), std::back_inserter(vias), [](const auto &alias) {
+                  const auto roomAliasDelimiter = ":";
+                  return alias.substr(alias.find(roomAliasDelimiter) + 1);
+          });
+
+        return vias;
+}
+
+void
+RoomDirectoryModel::joinRoom(const int &index)
+{
+        constexpr int delegateNoLongerValid = -1;
+        if (index == delegateNoLongerValid)
+                return;
+
+        const auto &chunk = publicRoomsData_[index];
+        nhlog::ui()->debug("'Joining room {}", chunk.room_id);
+        ChatPage::instance()->joinRoomVia(chunk.room_id, getViasForRoom(chunk.aliases));
+}
+
+QVariant
+RoomDirectoryModel::data(const QModelIndex &index, int role) const
+{
+        if (hasIndex(index.row(), index.column(), index.parent())) {
+                const auto &room_chunk = publicRoomsData_[index.row()];
+                switch (role) {
+                case Roles::Name:
+                        return QString::fromStdString(room_chunk.name);
+                case Roles::Id:
+                        return QString::fromStdString(room_chunk.room_id);
+                case Roles::AvatarUrl:
+                        return QString::fromStdString(room_chunk.avatar_url);
+                case Roles::Topic:
+                        return QString::fromStdString(room_chunk.topic);
+                case Roles::MemberCount:
+                        return QVariant::fromValue(room_chunk.num_joined_members);
+                case Roles::Previewable:
+                        return QVariant::fromValue(room_chunk.world_readable);
+                }
+        }
+        return {};
+}
+
+void
+RoomDirectoryModel::fetchMore(const QModelIndex &)
+{
+        nhlog::net()->debug("Fetching more rooms from mtxclient...");
+        nhlog::net()->debug("Prev batch: {} | Next batch: {}", prevBatch_, nextBatch_);
+
+        mtx::requests::PublicRooms req;
+        req.limit                      = limit_;
+        req.since                      = prevBatch_;
+        req.filter.generic_search_term = userSearchString_;
+        // req.third_party_instance_id = third_party_instance_id;
+
+        http::client()->post_public_rooms(
+          req,
+          [this, req](const mtx::responses::PublicRooms &res, mtx::http::RequestErr err) {
```

Comment:
You might want to check, that once this request completes searchTerm and since are still the same as is the searched server.